### PR TITLE
feature: make starting in-flight exit input transaction type agnostic

### LINKED ISSUE #205: Start Inflight Exit should be input tx type agnostic
### What
In start IFE PR: https://github.com/omisego/plasma-contracts/pull/196, it directly uses `PaymentTransaction` as input transaction. We would like to not hardcode input tx in the `PaymentInflightExitable`.

### Why
We want to be able to **not** audit again the main code when we are adding DEX feature or extend to any other tx that can be input of `Payment transaction`.

### Proposal solution
We could probably have a `verifyStateTransition` predicate for IFE. This would basically check given input txs and the tx with the proof/witness, the state transition in such transaction is valid.

We would swap the predicate when we are changing/adding new input txs. Dependency tx types can be hardcoded there.

### Successful Critieria
- The `PaymentInflightExitable` should not hardcode `PaymentOuput` as input. (one example [here](https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/contracts/src/exits/payment/PaymentInFlightExitable.sol#L243), should be more in current code base)
- In `PaymentExitDataModel`, `inputs` of `InflightExit` should not be `PaymentOutput`: [code here](https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol#L34). (I guess need to be sth like `bytes`, or if there is better idea?)

### PR DESCRIPTION
**Pull request by pgebal**
Closes #205

### CODE REVIEW

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol [hunk 1]
```diff
@@ -212,18 +213,18 @@ library PaymentStartInFlightExit {
...
+                );
         }
     }
 
-    function verifyInputsSpendingCondition(StartExitData memory exitData) private view {
+    function verifyInputsSpent(StartExitData memory exitData) private view {
         for (uint i = 0; i < exitData.inputTxs.length; i++) {
             uint16 outputIndex = exitData.inputUtxosPos[i].outputIndex();
-            bytes32 outputGuard = exitData.inputTxs[i].outputs[outputIndex].outputGuard;
+            WireTransaction.Output memory output = WireTransaction.getOutput(exitData.inputTxs[i], outputIndex);
```
Comment (pgebal) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: Here a "cheat" is done.
Objective of the PR is to abstract from input transaction types. 
This line assumes, that we can extract some "output" data from transaction knowing only that it conforms to wire transaction format (as in Wire Transaction Format Proposal doc).
It would be very handy and make life easier, if all our transaction conforms to that format.
If we don't want to assume that. We could use an interface here, like we do with state transition verifier.

File: plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol [hunk 2]
```diff
@@ -25,6 +25,13 @@ library PaymentExitDataModel {
         mapping (uint192 => PaymentExitDataModel.StandardExit) exits;
     }
 
+    struct WithdrawData {
+        bytes32 outputGuard;
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol]: This must be `exitTarget` or, something represents the address to withdraw.

File: plasma_framework/contracts/mocks/exits/payment/routers/PaymentInFlightExitRouterMock.sol [hunk 3]
```diff
@@ -19,11 +20,11 @@ contract PaymentInFlightExitRouterMock is PaymentInFlightExitRouter {
         inFlightExitMap.exits[exitId].exitMap = Bits.setBit(inFlightExitMap.exits[exitId].exitMap, 255);
     }
 
-    function getInFlightExitInput(uint192 exitId, uint8 inputIndex) public view returns (PaymentOutputModel.Output memory) {
+    function getInFlightExitInput(uint192 exitId, uint8 inputIndex) public view returns (PaymentExitDataModel.WithdrawData memory) {
```
Comment (boolafish) [inline plasma_framework/contracts/mocks/exits/payment/routers/PaymentInFlightExitRouterMock.sol]: minor: for `inputIndex ` and `outputIndex` below, probably bump it to uint16 to be consistent with our current index size.

File: plasma_framework/contracts/src/exits/IStateTransitionVerifier.sol [hunk 4]
```diff
@@ -0,0 +1,17 @@
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+interface IStateTransitionVerifier {
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/IStateTransitionVerifier.sol]: I would put this under `.../src/exits/interfaces/IStateTransitionVerifier.sol`. I think we will have a few interfaces for to be put there: `IOutputGuardHandler` (#227 ), `ISpendingCondition`(#214 ) and this one (and potentially more).

File: plasma_framework/contracts/src/transactions/WireTransaction.sol [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    using RLP for RLP.RLPItem;
+
+    struct Output {
+        uint256 amount;
+        bytes32 outputGuard;
+        address token;
+    }
+
+    /**
+    * @notice Returns output for transaction in wire format.
```
Comment (boolafish) [inline plasma_framework/contracts/src/transactions/WireTransaction.sol]: I would add the assumption of using this function in @dev here.
Like the `outputs` must be a field on the second index and should be a list which is the format of {amount, outputGuard, token} accordingly.

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol [hunk 6]
```diff
@@ -309,14 +270,39 @@ library PaymentStartInFlightExit {
...
+    }
+
+    function setInFlightExitOutputs(
+        PaymentExitDataModel.InFlightExit storage ife,
+        StartExitData memory exitData
+    )
+        private
+    {
+        for (uint i = 0; i < exitData.inFlightTx.outputs.length; i++) {
+            bytes32 outputId = OutputId.computeNormalOutputId(exitData.inFlightTxRaw, i);
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: I would add a comment that deposit tx cannot in-flight exit.

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol [hunk 7]
```diff
@@ -309,14 +270,39 @@ library PaymentStartInFlightExit {
...
+            ife.inputs[i] = PaymentExitDataModel.WithdrawData(
+                output.outputGuard,
+                exitData.outputIds[i],
+                output.token,
+                output.amount
+            );
+        }
+    }
+
+    function setInFlightExitOutputs(
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: haha okay, if we are setting outputs here, I will think again whether I want to separate the piggyback into two functions or not in PR #222 . I personal have preference over simplicity than optimization so I do like to set outputs here to make logic consistent. Let's see how the gas goes.

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol (hunk 1 above)
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: hmm, with #214 the new interface probably wouldn't need to getOutput here. However, we do need at least some way to get the `WithdrawData` in `setInputs(...)`. 

For DEX related txs as input tx this assumption seems fine as all tx outputs would be using same format anyway. For `ERC721` I think it *can* work because we are only getting output here, but I doubt the tx format would be as described in the document. I feel when we introduce atomic swap tx for fungible and non fungible token, it would be more or less like the following:

```
{
    txType: int.
    inputs: [bytes32, ...],
    fungibleTokenOutputs: [{ }, { },...],
    nonfungibleTokenOutputs: [{ }, { }, ...],
    // other fields
}
```

Reason of two output lists is that they need different data structure. I don't think current {amount, outputGuard, token} could fit non-fungible token output. And, at least in solidity, you cannot have one outputs array with multiple struct within it.

So, yes, it *can* work as long as we put the fungible token outputs in the second indexed position. However, whether it is a good idea to do would be more controversial. I think it strongly depends on what `WireTransaction` means and what assumption we are talking about when we are introducing such tx format.

File: plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            inputs[i] = output;
+        }
+
+        WireTransaction.Output[] memory outputs = new WireTransaction.Output[](inputTxs.length);
+        PaymentTransactionModel.Transaction memory transaction = PaymentTransactionModel.decode(inFlightTx);
+        for (uint i = 0; i < transaction.outputs.length; i++) {
+            outputs[i] = WireTransaction.Output(transaction.outputs[i].amount, transaction.outputs[i].outputGuard, transaction.outputs[i].token);
+        }
+
+        return _isCorrectStateTransition(inputs, outputs);
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol]: Since you're reverting on length mismatch earlier, probably we can consider reverting only strategy to keep consistent or we return false on the earlier. I am on one hand not too much a fan of this reverting strategy as the boolean value become weird, but on the other hand reverting here provides better customized revert message. Also, for unknown reason, revert + return true at the end seems to be the common practice in smart contract (?) 

Anyway, I will leave this to your decision.

File: plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            if (outputs[i].token == token) {
+                outputsWithToken[j] = outputs[i];
+                j += 1;
+            }
+        }
+
+        return outputsWithToken;
+    }
+
+    function isCorrectSpend(
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol]: Minor: would `isCorrectSumRelationship` better describe the function?

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol [hunk 10]
```diff
@@ -101,18 +111,18 @@ library PaymentStartInFlightExit {
...
         exitData.inputTxsInclusionProofs = args.inputTxsInclusionProofs;
         exitData.inFlightTxWitnesses = args.inFlightTxWitnesses;
-        exitData.outputIds = getOutputIds(controller, exitData.inputTxsRaw, exitData.inputUtxosPos);
+        exitData.outputIds = getOutputIds(controller, exitData.inputTxs, exitData.inputUtxosPos);
         return exitData;
     }
 
     function decodeInputTxsPositions(uint256[] memory inputUtxosPos) private pure returns (UtxoPosLib.UtxoPos[] memory) {
-        require(inputUtxosPos.length <= MAX_INPUT_NUM, "To many input transactions provided");
+        require(inputUtxosPos.length <= MAX_INPUT_NUM, "To many transactions provided");
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: typo: "Too" many....

File: plasma_framework/test/src/transactions/WireTransaction.test.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        const encoded = web3.utils.bytesToHex(transaction.rlpEncoded());
+
+        const actual = await this.test.getOutput(encoded, 0);
+
+        expect(actual.outputGuard).to.equal(OUTPUT_GUARD);
+        expect(new BN(actual.amount)).to.be.bignumber.equal(new BN(AMOUNT));
+        expect(actual.token).to.equal(constants.ZERO_ADDRESS);
+    });
+
+    it('should decode custom transaction output', async () => {
```
Comment (boolafish) [inline plasma_framework/test/src/transactions/WireTransaction.test.js]:
```suggestion
    it('should decode custom transaction output that fulfills wire transaction format', async () => {
```

File: plasma_framework/test/src/exits/payment/PaymentStartInFlightExit.test.js [hunk 12]
```diff
@@ -429,11 +431,24 @@ contract('PaymentInFlightExitRouter', ([_, alice, bob, carol]) => {
...
+                    MIN_EXIT_PERIOD, DUMMY_INITIAL_IMMUNE_VAULTS_NUM, INITIAL_IMMUNE_EXIT_GAME_NUM,
+                );
+                this.exitGame = await PaymentInFlightExitRouter.new(
+                    this.framework.address,
+                    this.spendingConditionRegistry.address,
+                    this.stateTransitionVerifierReject.address,
+                );
+            });
+
+            it('should fail when in-flight transaction is an invalid state transistion', async () => {
```
Comment (boolafish) [inline plasma_framework/test/src/exits/payment/PaymentStartInFlightExit.test.js]:
```suggestion
            it('should fail when the in-flight transaction has an invalid state transistion', async () => {
```

Comment (boolafish) [reply plasma_framework/test/src/exits/payment/PaymentStartInFlightExit.test.js]: I might try to test both verifier returns false and reverts. I think both are potential results in such verifier.

File: plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol (hunk 2 above)
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol]: leave some discussion from slack:

rambo.pawel Today at 4:47 PM
@boolafish I think we can't store beneficiaries of exits when starting IFE, because that means we would have to provide output guard data to recover output owners from output guard, and that's private and may not be known by someone starting ife. We need to go with storing output gaurd, and then interested users provide output guard data on piggybacks @pepesza, is that correct? Maybe I am missing something about that privacy. (edited)
2 replies

rambo.pawel  < 1 minute ago
I talked to Andy, and output guard data is available for input transactions (already mined, included in chain), and may not be available for in-flight transaction itself. So we'll require to provide output guard data on piggyback.

rambo.pawel  < 1 minute ago
when starting IFE output guard data for input transactions is still needed for checking signatures (edited)

pepesza  1 minute ago
Yes, when owner is piggybacking, she needs to reveal the output guard data. This is a moment we fill in the blanks

pepesza  < 1 minute ago
Also, this mechanism only provides some privacy for funds at rest. When funds are being spend or are exited - data is being made public

---
Some notes agreement:
1. startIFE passes in the output guard pre-image. However, only used to check confirm sig (and would not be implemented in this PR). Beneficiaries is not checked at that time.
1. piggyback (input/output) checks the real beneficiaries of exit from output guard preimage and store it. Though input *can* do that while startIFE, move it to piggyback to make piggyback input/output logic more consistent.

File: plasma_framework/test/src/exits/payment/PaymentStartInFlightExit.test.js (hunk 12 above)
Comment (pgebal) [reply plasma_framework/test/src/exits/payment/PaymentStartInFlightExit.test.js]: done

File: plasma_framework/test/src/transactions/WireTransaction.test.js (hunk 11 above)
Comment (pgebal) [reply plasma_framework/test/src/transactions/WireTransaction.test.js]: done

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol (hunk 10 above)
Comment (pgebal) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: done

File: plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol (hunk 9 above)
Comment (pgebal) [reply plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol]: done

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol (hunk 6 above)
Comment (pgebal) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: done

File: plasma_framework/contracts/src/transactions/WireTransaction.sol (hunk 5 above)
Comment (pgebal) [reply plasma_framework/contracts/src/transactions/WireTransaction.sol]: done

File: plasma_framework/contracts/src/exits/IStateTransitionVerifier.sol (hunk 4 above)
Comment (pgebal) [reply plasma_framework/contracts/src/exits/IStateTransitionVerifier.sol]: done

File: plasma_framework/contracts/mocks/exits/payment/routers/PaymentInFlightExitRouterMock.sol (hunk 3 above)
Comment (pgebal) [reply plasma_framework/contracts/mocks/exits/payment/routers/PaymentInFlightExitRouterMock.sol]: done

File: plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol (hunk 2 above)
Comment (pgebal) [reply plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol]: done

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol (hunk 1 above)
Comment (pgebal) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: Let's wait for clarification. We'll update that if needed. As of now I would leave it like it is, as it conforms to Wire Transaction Format doc.

File: plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol (hunk 8 above)
Comment (pgebal) [reply plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol]: Done. I went for always returning a boolean. It's more flexible (but on the other hand what can a caller do if it returns false?, will probably always reverse).

File: plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol (hunk 2 above)
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol]: sigh) I end up realize it is such a hassle to get output guard from unknown input tx. So I think probably we should only do such complex thing once. In my piggyback PR I end up asking for an extra array of outputGuards for input. 

like to discuss with you on this: https://github.com/omisego/plasma-contracts/pull/222/files#r318546401

File: plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol (hunk 9 above)
Comment (pgebal) [reply plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol]: done

File: plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol (hunk 2 above)
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/PaymentExitDataModel.sol]: As we agree on having `exitTarget` set during inputs, let's sync on the data structure of our PRs so we can decrease merging conflicts. Below is my proposal:

```
    struct WithdrawData {
         bytes32 outputId;
         address payable exitTarget;
         address token;
         uint256 amount;
     }
```

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol (hunk 1 above)
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: sure!

File: plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol (hunk 9 above)
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/PaymentTransactionStateTransitionVerifier.sol]: haha not really done XD but this is minor anyway

**Comment by boolafish** [comment]
For some reason I am still awake lol. The change for output guard is outdated. Please check the newer code with outputguardHandler

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol (hunk 1 above)
Comment (boolafish) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: Let's just take in WireTransaction as it is for this PR. I am just noting down more concerns with current approach.

During the call with @paulperegud and @pdobacz yesterday, we are talking about that even the `output` for a DEX settlement might not follow this original structure, where we have to make `outputType` explicit on the output itself. So sth like `(outputType, outputGuard/owner, amount, token)`.

TL;DR we might need to have 2 kinds of output in DEX where one send all funds back to user and another keep fund in exchange due to the attack scenerio described [here](https://github.com/omisego/docs/blob/0773484b2a16872e8c730a4f8648d1dcb317a9b1/docs/restricted_custody_mvp1_spec.md#alice-wants-to-withdraw-an-initial-deposit-from-an-offline-venue-but-settlement-occurs). To be able to differentiate which mechanism to parse the `outputGuard/owner` as ch-ch would need to check when collecting txs, we need a flag thus we need the `outputType` being explicit.

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol [hunk 13]
```diff
@@ -309,14 +284,50 @@ library PaymentStartInFlightExit {
...
     {
-        for (uint i = 0; i < inputTxs.length; i++) {
-            uint16 outputIndex = inputUtxosPos[i].outputIndex();
-            ife.inputs[i] = inputTxs[i].outputs[outputIndex];
+        for (uint i = 0; i < exitData.inputTxs.length; i++) {
+            uint16 outputIndex = exitData.inputUtxosPos[i].outputIndex();
+            WireTransaction.Output memory output = WireTransaction.getOutput(exitData.inputTxs[i], outputIndex);
+
+            address payable exitTarget;
+            if (exitData.inputUtxosTypes[i] == 0) {
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: You'll need to change this when merging with master. As I stated in another comment we changed the way of handling outputGuard

File: plasma_framework/contracts/src/exits/payment/routers/PaymentInFlightExitRouterArgs.sol [hunk 14]
```diff
@@ -15,6 +16,7 @@ library PaymentInFlightExitRouterArgs {
         bytes[] inputTxs;
         uint256[] inputUtxosPos;
         uint256[] inputUtxosTypes;
+        bytes[] outputGuardDataPreImages;
```
Comment (boolafish) [inline plasma_framework/contracts/src/exits/payment/routers/PaymentInFlightExitRouterArgs.sol]: Minor: I usually use `outputGuardPreimage` in other places. But I will leave this to you.

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol (hunk 13 above)
Comment (pgebal) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentStartInFlightExit.sol]: done
